### ISSUE TITLE
func_webhookstash won't stash file larger than 100k, just hangs

### ISSUE THREAD

**Issue by nkirsch** [alexellis/faas#31]
I'm able to stash (via func_webhookstash) a small text file, but when attempting a larger file (1.8M) the command simply hangs.

## Expected Behavior
I expect no behavior difference, regardless of file size/type.

Or I expect a meaningful error.
## Current Behavior
The larger upload hangs indefinitely.

I do see the gateway forwarding to webhookstash:
```
docker logs -f func_gateway.1.mazbcz33vjo5lhkviyj2y5kb3  

Resolving: 'func_webhookstash'
[1490657353] Forwarding request [application/x-www-form-urlencoded] to: http://10.0.0.20:8080/
```

However, I see no further logging information.

## Steps to Reproduce (for bugs)
```
dev % ls -al hello.txt IMG_0733.JPG
-rw-r--r-- 1 nick nick      12 Mar 27 16:22 hello.txt
-rw-r--r-- 1 nick nick 1815980 Mar 27 13:44 IMG_0733.JPG

dev % cat hello.txt | curl -X POST http://localhost:8080/function/func_webhookstash --data-binary @-  
Stashing request

dev % cat IMG_0733.JPG | curl -X POST http://localhost:8080/function/func_webhookstash --data-binary @-
```
**hangs indefinitely**

## Environment

* Docker version 17.03.0-ce, build 60ccb22
* Linux dev 4.8.0-41-generic #44-Ubuntu SMP Fri Mar 3 15:27:17 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* Ubuntu 16.10

**Comment by nkirsch** [comment]
Seems to break at 100K:

```
dev % dd if=/dev/zero of=output.zero bs=1K count=50                                                                                                                                                                                                                      
50+0 records in
50+0 records out
51200 bytes (51 kB, 50 KiB) copied, 0.000235105 s, 218 MB/s

dev % cat output.zero | curl -X POST http://localhost:8080/function/func_webhookstash --data-binary @-
Stashing request

dev % dd if=/dev/zero of=output.zero bs=1K count=100
100+0 records in
100+0 records out
102400 bytes (102 kB, 100 KiB) copied, 0.000343563 s, 298 MB/s

dev % cat output.zero | curl -X POST http://localhost:8080/function/func_webhookstash --data-binary @-    
```

**Comment by nkirsch** [comment]
Note this also fails using the `-d @IMG_0733.JPG` method (vs `cat` and `-d @-`)

**Comment by alexellis** [comment]
I'll have a look into this issue, but a web-hook is just an event - not a blob. I.e. a JSON POST rather than binary images or dd-generated etc. It should be a short message in JSON plain-text or application/json.

**Comment by alexellis** [comment]
This could be related to the way `curl` tries to upload the file.

During debugging I saw `curl` uploading a small JPG with `x-www-form-urlencoded` content-type, that was transmitted and written out fine. If the size is larger i.e. above around 65600 bytes it appears to choke. Larger files are not meant to be uploaded with `x-www-form-urlencoded`.

The content-type `x-www-form-urlencoded` basically creates a huge query-string which may be getting blocked by the default Golang settings. A multipart form is probably the way to go when uploading large binary data and this may need a tweak to the watchdog process.

http://stackoverflow.com/questions/4007969/application-x-www-form-urlencoded-or-multipart-form-data

**Comment by alexellis** [comment]
Looks like the Golang buffered pipe in (watchdog process) was blocking at 64kb.

This is the fix - https://github.com/alexellis/faas/compare/watchdog_buffer_fix

If you'd like to try this before it's released then checkout the watchdog_buffer_fix branch and re-build the fwatchdog binary with the "build.sh" script in the watchdog folder. Then copy it to the webhook stash function.

**Comment by alexellis** [comment]
Fixed by #33

**Comment by nkirsch** [comment]
Thanks, @alexellis !
